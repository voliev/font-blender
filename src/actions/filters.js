import {
  SET_CATEGORIES_FILTER,
  SET_SORTING_ORDER,
  SET_SORTING_PARAM,
  SET_TITLE_FILTER
} from "../constants";

export const setTitleFilter = (title = "") => ({
  type: SET_TITLE_FILTER,
  title
});

export const setCategoriesFilter = category => ({
  type: SET_CATEGORIES_FILTER,
  category
});

export const setSortingParam = param => ({
  type: SET_SORTING_PARAM,
  param
});

export const setSortingOrder = order => ({
  type: SET_SORTING_ORDER,
  order
});
