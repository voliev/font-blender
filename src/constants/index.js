// *** ACTION TYPES ***

// Background
export const SET_BACKGROUND_COLOR = "SET_BACKGROUND_COLOR";

// Filteres
export const SET_CATEGORIES_FILTER = "SET_CATEGORIES_FILTER";
export const SET_SORTING_ORDER = "SET_SORTING_ORDER";
export const SET_SORTING_PARAM = "SET_SORTING_PARAM";
export const SET_TITLE_FILTER = "SET_TITLE_FILTER";

// Fonts
export const FETCH_FONTS_DATA = "FETCH_FONTS_DATA";
export const FETCH_FONTS_DATA_SUCCESS = "FETCH_FONTS_DATA_SUCCESS";
export const FETCH_FONTS_DATA_FAILURE = "FETCH_FONTS_DATA_FAILURE";
export const SET_FONTS = "SET_FONTS";
export const UPDATE_FONT = "UPDATE_FONT";

// Notifications
export const ADD_NOTIFICATION = "ADD_NOTIFICATION";
export const REMOVE_NOTIFICATION = "REMOVE_NOTIFICATION";

// Showcase
export const DISPLAY_SHOWCASE_TEXT = "DISPLAY_SHOWCASE_TEXT";
export const DISPLAY_STYLES = "DISPLAY_STYLES";
export const DISPLAY_SHOWCASE_FONT_PREVIEW = "DISPLAY_SHOWCASE_FONT_PREVIEW";

// Styles
export const SET_DEFAULT_STATE = "SET_DEFAULT_STATE";
export const UNDO_RESET = "UNDO_RESET";

// Text styles
export const SET_STYLES_INHERITANCE = "SET_STYLES_INHERITANCE";
export const UPDATE_ELEMENT_STYLES = "UPDATE_ELEMENT_STYLES";

// Selected fonts
export const ADD_SELECTED_FONT = "ADD_SELECTED_FONT";
export const CLEAR_SELECTED_FONTS = "CLEAR_SELECTED_FONTS";
export const REMOVE_FONT_FROM_SELECTED = "REMOVE_FONT_FROM_SELECTED";
export const UNDO_DESELECTED = "UNDO_DESELECTED";

// Type scale
export const SET_DEFAULT_TYPE_SCALE_STATE = "SET_DEFAULT_TYPE_SCALE_STATE";
export const SET_FONT_SIZE = "SET_FONT_SIZE";
export const SET_TYPE_SCALE = "SET_TYPE_SCALE";

// *** MISCELLANEOUS ***

// Google Fonts API endpoint URL
export const GOOGLE_FONTS_API_ENDPOINT =
  "https://www.googleapis.com/webfonts/v1/webfonts";
// Fonts' bundle size that should be prefetched
export const BUNDLE_SIZE = 50;
// Available text elements for styling
export const ELEMENTS = ["body", "h1", "h2", "h3", "h4", "li", "small"];
// Font's character range that needs to be prefetched
export const CHAR_RANGE =
  "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
// Font showcase string
export const SHOWCASE_STRING =
  "That's the beauty of music. They can't get that from you...";
// Default showcase font size
export const DEFAULT_FONT_SIZE = 16;
// Available type scale values
export const SCALE_VALUES = {
  1.067: "Minor Second",
  1.125: "Major Second",
  1.2: "Minor Third",
  1.25: "Major Third",
  1.333: "Perfect Fourth",
  1.414: "Augmented Fourth",
  1.5: "Perfect Fifth",
  1.618: "Golden Ratio"
};
// Minimum gap between the color picker icon and viewport
// edge required to place the color picker
export const MIN_GAP = 300;
// Notification snackbar visibility timeout
export const TIMEOUT = 10000;
