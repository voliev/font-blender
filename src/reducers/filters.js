import {
  SET_CATEGORIES_FILTER,
  SET_TITLE_FILTER,
  SET_SORTING_ORDER,
  SET_SORTING_PARAM
} from "../constants";

const initialState = {
  title: "",
  categories: {
    serif: true,
    "sans-serif": true,
    display: true,
    monospace: true,
    handwriting: true
  },
  sorting: {
    param: "popularity",
    order: "asc"
  }
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_TITLE_FILTER:
      return {
        ...state,
        title: action.title
      };
    case SET_CATEGORIES_FILTER:
      return {
        ...state,
        categories: {
          ...state.categories,
          [action.category]: !state.categories[action.category]
        }
      };
    case SET_SORTING_PARAM:
      return {
        ...state,
        sorting: {
          ...state.sorting,
          param: action.param
        }
      };
    case SET_SORTING_ORDER:
      return {
        ...state,
        sorting: {
          ...state.sorting,
          order: action.order === "asc" ? "des" : "asc"
        }
      };
    default:
      return state;
  }
};

export const getFilters = state => state.filters;
