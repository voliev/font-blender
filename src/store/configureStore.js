import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import fonts from "../reducers/fonts";
import filters from "../reducers/filters";
import selectedFonts from "../reducers/selectedFonts";
import showcase from "../reducers/showcase";
import styles from "../reducers/styles";
import notifications from "../reducers/notifications";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default () => {
  const store = createStore(
    combineReducers({
      fonts,
      filters,
      selectedFonts,
      showcase,
      styles,
      notifications
    }),
    composeEnhancers(applyMiddleware(thunk))
  );

  return store;
};
