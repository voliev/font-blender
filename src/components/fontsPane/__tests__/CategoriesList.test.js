import React from "react";
import { mount, shallow } from "enzyme";
import configureStore from "redux-mock-store";
import CheckBoxOutlineBlank from "@material-ui/icons/CheckBoxOutlineBlank";
import CheckBoxOutlined from "@material-ui/icons/CheckBoxOutlined";
import CategoriesListConnected, { CategoriesList } from "../CategoriesList";
import { SET_CATEGORIES_FILTER } from "../../../constants";
import filters from "../../../fixtures/filters";

const setup = (renderFn = shallow, propsOverrides = {}, connected = false) => {
  const props = {
    categories: { ...filters.categories },
    stage: "fade-in",
    animationEndHandler: jest.fn(),
    setCategoriesFilter: jest.fn(),
    ...propsOverrides
  };

  if (connected) {
    const mockStore = configureStore();
    const store = mockStore({ filters });

    return {
      store,
      wrapper: shallow(
        <CategoriesListConnected
          animationEndHandler={props.animationEndHandler}
          stage={props.stage}
          store={store}
        />
      )
    };
  }

  return {
    ...props,
    wrapper: renderFn(<CategoriesList {...props} />)
  };
};

it("should render without crashing", () => {
  const { wrapper } = setup(mount);

  expect(wrapper).toMatchSnapshot();
});

it("should receive corresponding props", () => {
  const {
    animationEndHandler,
    categories,
    setCategoriesFilter,
    stage,
    wrapper
  } = setup(mount);

  expect(typeof wrapper.prop("categories")).toBe("object");
  expect(wrapper.prop("categories")).toEqual(categories);

  expect(typeof wrapper.prop("setCategoriesFilter")).toBe("function");
  expect(wrapper.prop("setCategoriesFilter")).toBe(setCategoriesFilter);

  expect(typeof wrapper.prop("animationEndHandler")).toBe("function");
  expect(wrapper.prop("animationEndHandler")).toBe(animationEndHandler);

  expect(typeof wrapper.prop("stage")).toBe("string");
  expect(wrapper.prop("stage")).toBe(stage);
});

it("should set corresponding component classes", () => {
  const { wrapper } = setup(mount);

  expect(wrapper.find("ul").hasClass("CategoriesList")).toBeTruthy();
  expect(wrapper.find("ul").hasClass("fade-in")).toBeTruthy();

  wrapper.setProps({
    stage: "active"
  });

  expect(wrapper.find("ul").hasClass("CategoriesList")).toBeTruthy();
  expect(wrapper.find("ul").hasClass("fade-in")).toBeFalsy();

  wrapper.setProps({
    stage: "fade-out"
  });

  expect(wrapper.find("ul").hasClass("CategoriesList")).toBeTruthy();
  expect(wrapper.find("ul").hasClass("fade-out")).toBeTruthy();
});

it("should handle input change", () => {
  const { wrapper, setCategoriesFilter } = setup(mount);
  const category = "serif";
  const event = {
    target: { value: category }
  };

  wrapper.find('input[value="serif"]').simulate("change", event);

  expect(setCategoriesFilter).toHaveBeenCalledTimes(1);
  expect(setCategoriesFilter).toHaveBeenCalledWith(category);
});

it("should dispatch an action", () => {
  const { store, wrapper } = setup(shallow, undefined, true);

  wrapper.prop("setCategoriesFilter")("serif");

  const actions = store.getActions();

  expect(actions).toContainEqual({
    type: SET_CATEGORIES_FILTER,
    category: "serif"
  });
});

it("should set corresponding checkbox icon", () => {
  const { wrapper, categories } = setup(mount);

  expect(wrapper.find(CheckBoxOutlined).length).toBe(5);
  expect(wrapper.find(CheckBoxOutlineBlank).length).toBe(0);

  wrapper.setProps({
    categories: {
      ...categories,
      serif: false
    }
  });

  expect(wrapper.find(CheckBoxOutlined).length).toBe(4);
  expect(wrapper.find(CheckBoxOutlineBlank).length).toBe(1);

  wrapper.setProps({
    categories: {
      ...categories,
      serif: false,
      "sans-serif": false,
      display: false,
      monospace: false,
      handwriting: false
    }
  });

  expect(wrapper.find(CheckBoxOutlined).length).toBe(0);
  expect(wrapper.find(CheckBoxOutlineBlank).length).toBe(5);
});
