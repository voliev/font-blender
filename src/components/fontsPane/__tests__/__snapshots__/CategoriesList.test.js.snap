// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
<CategoriesList
  animationEndHandler={[MockFunction]}
  categories={
    Object {
      "display": true,
      "handwriting": true,
      "monospace": true,
      "sans-serif": true,
      "serif": true,
    }
  }
  setCategoriesFilter={[MockFunction]}
  stage="fade-in"
>
  <ul
    className="CategoriesList fade-in"
    onAnimationEnd={[MockFunction]}
  >
    <li
      className="list-item"
      key="serif"
    >
      <label
        className="label"
        htmlFor="serif"
      >
        <ForwardRef>
          <WithStyles(ForwardRef(SvgIcon))>
            <ForwardRef(SvgIcon)
              classes={
                Object {
                  "colorAction": "MuiSvgIcon-colorAction",
                  "colorDisabled": "MuiSvgIcon-colorDisabled",
                  "colorError": "MuiSvgIcon-colorError",
                  "colorPrimary": "MuiSvgIcon-colorPrimary",
                  "colorSecondary": "MuiSvgIcon-colorSecondary",
                  "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                  "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                  "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                  "root": "MuiSvgIcon-root",
                }
              }
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V5h14v14zM17.99 9l-1.41-1.42-6.59 6.59-2.58-2.57-1.42 1.41 4 3.99z"
                />
              </svg>
            </ForwardRef(SvgIcon)>
          </WithStyles(ForwardRef(SvgIcon))>
        </ForwardRef>
        <input
          checked={true}
          className="input"
          id="serif"
          name="categories"
          onChange={[Function]}
          type="checkbox"
          value="serif"
        />
        Serif
      </label>
    </li>
    <li
      className="list-item"
      key="sans-serif"
    >
      <label
        className="label"
        htmlFor="sans-serif"
      >
        <ForwardRef>
          <WithStyles(ForwardRef(SvgIcon))>
            <ForwardRef(SvgIcon)
              classes={
                Object {
                  "colorAction": "MuiSvgIcon-colorAction",
                  "colorDisabled": "MuiSvgIcon-colorDisabled",
                  "colorError": "MuiSvgIcon-colorError",
                  "colorPrimary": "MuiSvgIcon-colorPrimary",
                  "colorSecondary": "MuiSvgIcon-colorSecondary",
                  "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                  "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                  "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                  "root": "MuiSvgIcon-root",
                }
              }
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V5h14v14zM17.99 9l-1.41-1.42-6.59 6.59-2.58-2.57-1.42 1.41 4 3.99z"
                />
              </svg>
            </ForwardRef(SvgIcon)>
          </WithStyles(ForwardRef(SvgIcon))>
        </ForwardRef>
        <input
          checked={true}
          className="input"
          id="sans-serif"
          name="categories"
          onChange={[Function]}
          type="checkbox"
          value="sans-serif"
        />
        Sans Serif
      </label>
    </li>
    <li
      className="list-item"
      key="display"
    >
      <label
        className="label"
        htmlFor="display"
      >
        <ForwardRef>
          <WithStyles(ForwardRef(SvgIcon))>
            <ForwardRef(SvgIcon)
              classes={
                Object {
                  "colorAction": "MuiSvgIcon-colorAction",
                  "colorDisabled": "MuiSvgIcon-colorDisabled",
                  "colorError": "MuiSvgIcon-colorError",
                  "colorPrimary": "MuiSvgIcon-colorPrimary",
                  "colorSecondary": "MuiSvgIcon-colorSecondary",
                  "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                  "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                  "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                  "root": "MuiSvgIcon-root",
                }
              }
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V5h14v14zM17.99 9l-1.41-1.42-6.59 6.59-2.58-2.57-1.42 1.41 4 3.99z"
                />
              </svg>
            </ForwardRef(SvgIcon)>
          </WithStyles(ForwardRef(SvgIcon))>
        </ForwardRef>
        <input
          checked={true}
          className="input"
          id="display"
          name="categories"
          onChange={[Function]}
          type="checkbox"
          value="display"
        />
        Display
      </label>
    </li>
    <li
      className="list-item"
      key="monospace"
    >
      <label
        className="label"
        htmlFor="monospace"
      >
        <ForwardRef>
          <WithStyles(ForwardRef(SvgIcon))>
            <ForwardRef(SvgIcon)
              classes={
                Object {
                  "colorAction": "MuiSvgIcon-colorAction",
                  "colorDisabled": "MuiSvgIcon-colorDisabled",
                  "colorError": "MuiSvgIcon-colorError",
                  "colorPrimary": "MuiSvgIcon-colorPrimary",
                  "colorSecondary": "MuiSvgIcon-colorSecondary",
                  "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                  "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                  "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                  "root": "MuiSvgIcon-root",
                }
              }
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V5h14v14zM17.99 9l-1.41-1.42-6.59 6.59-2.58-2.57-1.42 1.41 4 3.99z"
                />
              </svg>
            </ForwardRef(SvgIcon)>
          </WithStyles(ForwardRef(SvgIcon))>
        </ForwardRef>
        <input
          checked={true}
          className="input"
          id="monospace"
          name="categories"
          onChange={[Function]}
          type="checkbox"
          value="monospace"
        />
        Monospace
      </label>
    </li>
    <li
      className="list-item"
      key="handwriting"
    >
      <label
        className="label"
        htmlFor="handwriting"
      >
        <ForwardRef>
          <WithStyles(ForwardRef(SvgIcon))>
            <ForwardRef(SvgIcon)
              classes={
                Object {
                  "colorAction": "MuiSvgIcon-colorAction",
                  "colorDisabled": "MuiSvgIcon-colorDisabled",
                  "colorError": "MuiSvgIcon-colorError",
                  "colorPrimary": "MuiSvgIcon-colorPrimary",
                  "colorSecondary": "MuiSvgIcon-colorSecondary",
                  "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                  "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                  "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                  "root": "MuiSvgIcon-root",
                }
              }
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V5h14v14zM17.99 9l-1.41-1.42-6.59 6.59-2.58-2.57-1.42 1.41 4 3.99z"
                />
              </svg>
            </ForwardRef(SvgIcon)>
          </WithStyles(ForwardRef(SvgIcon))>
        </ForwardRef>
        <input
          checked={true}
          className="input"
          id="handwriting"
          name="categories"
          onChange={[Function]}
          type="checkbox"
          value="handwriting"
        />
        Handwriting
      </label>
    </li>
  </ul>
</CategoriesList>
`;
